src/VisualStudio/VisualBasic/Impl/LanguageService/VisualBasicPackage.vb

```diff
@@ -84,40 +84,47 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic
             Return Me.ComponentModel.GetService(Of VisualStudioWorkspaceImpl)
         End Function
 
-        Protected Overrides Sub Initialize()
+        Protected Overrides Async Function InitializeAsync(cancellationToken As CancellationToken, progress As IProgress(Of ServiceProgressData)) As Task
             Try
-                MyBase.Initialize()
+                Await MyBase.InitializeAsync(cancellationToken, progress).ConfigureAwait(True)
+                Await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken)
 
                 RegisterLanguageService(GetType(IVbCompilerService), Function() _comAggregate)
 
                 Dim workspace = Me.ComponentModel.GetService(Of VisualStudioWorkspaceImpl)()
                 RegisterService(Of IVbTempPECompilerFactory)(Function() New TempPECompilerFactory(workspace))
 
-                RegisterObjectBrowserLibraryManager()
+                Await RegisterObjectBrowserLibraryManagerAsync(cancellationToken).ConfigureAwait(True)
             Catch ex As Exception When FatalError.Report(ex)
             End Try
-        End Sub
+        End Function
 
         Protected Overrides Sub Dispose(disposing As Boolean)
-            UnregisterObjectBrowserLibraryManager()
+            If disposing Then
+                JoinableTaskFactory.Run(Function() UnregisterObjectBrowserLibraryManagerAsync(CancellationToken.None))
+            End If
 
             MyBase.Dispose(disposing)
         End Sub
 
-        Private Sub RegisterObjectBrowserLibraryManager()
-            Dim objectManager = TryCast(Me.GetService(GetType(SVsObjectManager)), IVsObjectManager2)
+        Private Async Function RegisterObjectBrowserLibraryManagerAsync(cancellationToken As CancellationToken) As Task
+            Dim objectManager = TryCast(Await GetServiceAsync(GetType(SVsObjectManager)).ConfigureAwait(True), IVsObjectManager2)
             If objectManager IsNot Nothing Then
+                Await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken)
+
                 Me._libraryManager = New ObjectBrowserLibraryManager(Me)
 
                 If ErrorHandler.Failed(objectManager.RegisterSimpleLibrary(Me._libraryManager, Me._libraryManagerCookie)) Then
                     Me._libraryManagerCookie = 0
                 End If
             End If
-        End Sub
+        End Function
 
-        Private Sub UnregisterObjectBrowserLibraryManager()
+        Private Async Function UnregisterObjectBrowserLibraryManagerAsync(cancellationToken As CancellationToken) As Task
             If _libraryManagerCookie <> 0 Then
-                Dim objectManager = TryCast(Me.GetService(GetType(SVsObjectManager)), IVsObjectManager2)
+                Await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken)
```

Comment:
this is exactly same as C# implementation. why do we have duplicated code?

Reply:
I am not aware of a specific reason, and could not find one in the comments or git blame. However, a change in this area currently seems outside the scope of this pull request.